// llamaindex@0.3.13/dist/cjs/storage/vectorStore/types.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    VectorStoreBase: function() {
        return VectorStoreBase;
    },
    VectorStoreQueryMode: function() {
        return VectorStoreQueryMode;
    }
});
const _EmbedModel = require("../../internal/settings/EmbedModel.js");
var VectorStoreQueryMode;
(function(VectorStoreQueryMode) {
    VectorStoreQueryMode["DEFAULT"] = "default";
    VectorStoreQueryMode["SPARSE"] = "sparse";
    VectorStoreQueryMode["HYBRID"] = "hybrid";
    // fit learners
    VectorStoreQueryMode["SVM"] = "svm";
    VectorStoreQueryMode["LOGISTIC_REGRESSION"] = "logistic_regression";
    VectorStoreQueryMode["LINEAR_REGRESSION"] = "linear_regression";
    // maximum marginal relevance
    VectorStoreQueryMode["MMR"] = "mmr";
})(VectorStoreQueryMode || (VectorStoreQueryMode = {}));
class VectorStoreBase {
    embedModel;
    constructor(embedModel){
        this.embedModel = embedModel ?? (0, _EmbedModel.getEmbeddedModel)();
    }
}
