// @datastax/astra-db-ts@1.1.0/dist/api/constants.js
"use strict";
// Copyright DataStax, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_DATA_API_PATH = exports.DEFAULT_DEVOPS_API_ENDPOINT = exports.DEFAULT_DEVOPS_API_AUTH_HEADER = exports.DEFAULT_DATA_API_AUTH_HEADER = exports.DEFAULT_TIMEOUT = exports.DEFAULT_NAMESPACE = exports.HttpMethods = exports.CLIENT_USER_AGENT = exports.RAGSTACK_REQUESTED_WITH = void 0;
const version_1 = require("../version");
/**
 * @internal
 */
exports.RAGSTACK_REQUESTED_WITH = (() => {
    try {
        // Do not use require() here, it will break the build in some environments such as NextJS application
        // if @datastax/ragstack-ai is not installed (which is perfectly fine).
        const ragstack = eval(`require('@datastax/ragstack-ai')`);
        const version = ragstack['RAGSTACK_VERSION'] || '?';
        return `ragstack-ai-ts/${version}`;
    }
    catch (e) {
        return '';
    }
})();
/**
 * @internal
 */
exports.CLIENT_USER_AGENT = version_1.LIB_NAME + '/' + version_1.LIB_VERSION;
/**
 * @internal
 */
exports.HttpMethods = {
    Get: 'GET',
    Post: 'POST',
    Delete: 'DELETE',
};
/**
 * @internal
 */
exports.DEFAULT_NAMESPACE = 'default_keyspace';
/**
 * @internal
 */
exports.DEFAULT_TIMEOUT = 30000;
/**
 * @internal
 */
exports.DEFAULT_DATA_API_AUTH_HEADER = 'Token';
/**
 * @internal
 */
exports.DEFAULT_DEVOPS_API_AUTH_HEADER = 'Authorization';
/**
 * @internal
 */
exports.DEFAULT_DEVOPS_API_ENDPOINT = 'https://api.astra.datastax.com/v2';
/**
 * @internal
 */
exports.DEFAULT_DATA_API_PATH = 'api/json/v1';
