// @datastax/astra-db-ts@1.1.0/dist/data-api/index.js
"use strict";
// Copyright DataStax, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// noinspection DuplicatedCode
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateManyError = exports.CursorIsStartedError = exports.BulkWriteError = exports.InsertManyError = exports.TooManyDocumentsToCountError = exports.DeleteManyError = exports.CollectionAlreadyExistsError = exports.DataAPIError = exports.CumulativeDataAPIError = exports.DataAPIResponseError = exports.DataAPITimeoutError = exports.Db = void 0;
__exportStar(require("./collection"), exports);
__exportStar(require("./cursor"), exports);
__exportStar(require("./document"), exports);
var db_1 = require("./db");
Object.defineProperty(exports, "Db", { enumerable: true, get: function () { return db_1.Db; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "DataAPITimeoutError", { enumerable: true, get: function () { return errors_1.DataAPITimeoutError; } });
Object.defineProperty(exports, "DataAPIResponseError", { enumerable: true, get: function () { return errors_1.DataAPIResponseError; } });
Object.defineProperty(exports, "CumulativeDataAPIError", { enumerable: true, get: function () { return errors_1.CumulativeDataAPIError; } });
Object.defineProperty(exports, "DataAPIError", { enumerable: true, get: function () { return errors_1.DataAPIError; } });
Object.defineProperty(exports, "CollectionAlreadyExistsError", { enumerable: true, get: function () { return errors_1.CollectionAlreadyExistsError; } });
Object.defineProperty(exports, "DeleteManyError", { enumerable: true, get: function () { return errors_1.DeleteManyError; } });
Object.defineProperty(exports, "TooManyDocumentsToCountError", { enumerable: true, get: function () { return errors_1.TooManyDocumentsToCountError; } });
Object.defineProperty(exports, "InsertManyError", { enumerable: true, get: function () { return errors_1.InsertManyError; } });
Object.defineProperty(exports, "BulkWriteError", { enumerable: true, get: function () { return errors_1.BulkWriteError; } });
Object.defineProperty(exports, "CursorIsStartedError", { enumerable: true, get: function () { return errors_1.CursorIsStartedError; } });
Object.defineProperty(exports, "UpdateManyError", { enumerable: true, get: function () { return errors_1.UpdateManyError; } });
__exportStar(require("./events"), exports);
__exportStar(require("./ids"), exports);
__exportStar(require("./types"), exports);
